certbot-ci/certbot_integration_tests/conftest.py

```diff
@@ -86,11 +93,26 @@ def _setup_primary_node(config):
     workers = ['primary'] if not config.option.numprocesses\
         else ['gw{0}'.format(i) for i in range(config.option.numprocesses)]
 
+    # If a non-default DNS server is configured, start it and feed it to the ACME server
+    dns_server = None
+    acme_dns_server = None
+    if config.option.dns_server == 'bind':
+        dns_server = dns_lib.DNSServer(workers)
+        config.add_cleanup(dns_server.stop)
+        print('DNS xdist config:\n{0}'.format(dns_server.dns_xdist))
+        dns_server.start()
+        acme_dns_server = '{}:{}'.format(
+            dns_server.dns_xdist['address'],
+            dns_server.dns_xdist[workers[0]]
+        )
+
     # By calling setup_acme_server we ensure that all necessary acme server instances will be
     # fully started. This runtime is reflected by the acme_xdist returned.
-    acme_server = acme_lib.ACMEServer(config.option.acme_server, workers)
+    acme_server = acme_lib.ACMEServer(config.option.acme_server, workers,
+                                      dns_server=acme_dns_server)
     config.add_cleanup(acme_server.stop)
```

Comment:
To be fair with my comment above about side effects, I did this kind of things here ... So if you agree with the general approach of avoiding side effects, maybe we could return the cleanup callback from `_setup_primary_node` and attach it to the cleanup mechanism afterwards.

Reply:
I agree with the sentiment generally. However, to do so by passing around a relatively complex tuple of things to inject,  seems more error-prone.

My thinking is that `pytest_configure_node` is a side-effect function by definition, I don't think it's so bad to continue along that path in the internal implementation.

Reply:
Yes, my focus on this code style may be influenced here by hard times I had troubleshooting bad behaviors happening in the pytest hook phase at the beginning. But now it is set up. I am ok with you, we can just mark `config` as mutated in the docstring of `_setup_primary_node`.